modules/flyte/roles.tf

```diff
@@ -1,30 +1,56 @@
-resource "azurerm_role_definition" "flyte_storage_access" {
-  name  = "${var.deploy_id}-flyte-storage-access"
+resource "azurerm_role_definition" "flyte_metadata_role" {
+  name  = "${var.deploy_id}-flyte-metadata"
   scope = azurerm_storage_container.flyte_metadata.resource_manager_id
   permissions {
-    actions = [
-      "Microsoft.Storage/storageAccounts/blobServices/containers/read",
-      "Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey/action"
+    data_actions = [
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read",
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write",
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/add/action",
     ]
+  }
+}
+
+resource "azurerm_role_definition" "flyte_data_role" {
+  name  = "${var.deploy_id}-flyte-data"
+  scope = azurerm_storage_container.flyte_data.resource_manager_id
+  permissions {
     data_actions = [
       "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read",
       "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write",
+      "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/add/action",
     ]
   }
-  assignable_scopes = [
-    azurerm_storage_container.flyte_metadata.resource_manager_id,
-    azurerm_storage_container.flyte_data.resource_manager_id,
-  ]
 }
 
-resource "azurerm_role_assignment" "flyte_metadata_access" {
+# Because the Get User Delegation Key operation acts at the level of the storage account, the
+# Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey action must be scoped at the level
+# of the storage account, the resource group, or the subscription.
+# https://learn.microsoft.com/en-us/rest/api/storageservices/get-user-delegation-key
+#
+resource "azurerm_role_definition" "flyte_sas_role" {
+  name  = "${var.deploy_id}-flyte-sas"
+  scope = azurerm_storage_account.flyte.id
+  permissions {
+    actions = [
+      "Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey/action"
+    ]
+  }
+}
+
+resource "azurerm_role_assignment" "flyte_metadata_role_assignment" {
```

Comment:
similar to previous comment, you can define a dict and just iterate in a single  `resource "azurerm_role_assignment"`

Reply:
Here is what I have with an `azurerm_role_assignment` dict:

```hcl
locals {
  role_assignment_mapping = {
    metadata = {
      scope              = azurerm_storage_container.flyte_metadata.resource_manager_id,
      role_definition_id = azurerm_role_definition.flyte_metadata.role_definition_resource_id,
    }
    data = {
      scope              = azurerm_storage_container.flyte_data.resource_manager_id,
      role_definition_id = azurerm_role_definition.flyte_data.role_definition_resource_id,
    }
    sas = {
      scope              = azurerm_storage_account.flyte.id,
      role_definition_id = azurerm_role_definition.flyte_sas.role_definition_resource_id,
    }
  }
}

resource "azurerm_role_assignment" "flyte" {
  for_each           = local.role_assignment_mapping
  scope              = each.value.scope
  role_definition_id = each.value.role_definition_id
  principal_id       = azuread_group.flyte[each.key].object_id
}
```

Maybe it's just me but I feel the new code is harder to comprehend with extra level of indirection compared to the original code where the resources are defined individually:

```hcl
resource "azurerm_role_assignment" "flyte_metadata" {
  scope              = azurerm_storage_container.flyte_metadata.resource_manager_id
  role_definition_id = azurerm_role_definition.flyte_metadata.role_definition_resource_id
  principal_id       = azuread_group.flyte_metadata.object_id
}

resource "azurerm_role_assignment" "flyte_data" {
  scope              = azurerm_storage_container.flyte_data.resource_manager_id
  role_definition_id = azurerm_role_definition.flyte_data.role_definition_resource_id
  principal_id       = azuread_group.flyte_data.object_id
}

resource "azurerm_role_assignment" "flyte_sas" {
  scope              = azurerm_storage_account.flyte.id
  role_definition_id = azurerm_role_definition.flyte_sas.role_definition_resource_id
  principal_id       = azuread_group.flyte_sas.object_id
}
```